fgpyo/sam/builder.py

```diff
@@ -116,14 +119,24 @@ def __init__(
             extra_header: a dictionary of extra values to add to the header, None otherwise.  See
                           `::class::~pysam.AlignmentHeader` for more details.
             seed: a seed value for random number/string generation
+            sort_order: optional sort order, if `None` reads will be output in the same order as
+                they were appended.
         """
 
         self.r1_len: int = r1_len if r1_len is not None else self.DEFAULT_R1_LENGTH
         self.r2_len: int = r2_len if r2_len is not None else self.DEFAULT_R2_LENGTH
         self.base_quality: int = base_quality
         self.mapping_quality: int = mapping_quality
+
+        sort_order = SamOrder.Unsorted if sort_order is None else sort_order
+        assert sort_order in [SamOrder.Coordinate, SamOrder.QueryName, SamOrder.Unsorted], (
+            "`sort_order` for `SamBuilder` must be one of `Coordinate` `QueryName` or `Unsorted`."
+            + f" Found {sort_order}"
+        )
+        self.sort_order: SamOrder = sort_order
```

Comment:
Can you propagate this as private so that folks don't mutate this and expect the header to change?
```suggestion
        self._sort_order: SamOrder = sort_order
```